Firestore/core/src/firebase/firestore/util/executor_std.h

```diff
@@ -93,8 +93,9 @@ class Schedule {
       // To minimize busy waiting, sleep until either the nearest entry in the
       // future either changes, or else becomes due.
       const auto until = scheduled_.front().due;
-      cv_.wait_until(lock, until,
-                     [this, until] { return scheduled_.front().due != until; });
+      cv_.wait_until(lock, until, [this, until] {
+        return scheduled_.empty() || scheduled_.front().due != until;
```

Comment:
I think the new condition should be inversed:
```cpp
return !scheduled_.empty() && scheduled_.front().due != until;
```
The condition variable waits until the given predicate becomes true. In essence, it's supposed to wait until the queue has an overdue task. So while waking up when `schedule_` becomes empty would actually work correctly (because `HasDueLocked` checks for emptiness), I think the intent would be better stated if this thread only wakes up when there is a potentially overdue task. I tried to think of a case when it would be important to wake up because the queue became empty, but couldn't think of one -- let me know if you have an example.

It's an interesting question why it works in libc++. By the standard, calling `front` on an empty `deque` is undefined behavior. libc++ doesn't have an assertion in their implementation of `front`. I haven't debugged it too much, but it's interesting that this access doesn't trigger ASan. Perhaps it hits a case which only MSan can see.

Reply:
The last question is the easiest: unless the backing store for the deque is deallocated after the last item is removed (and I'm not aware of any implementation that would do that) the backing array will still exist, and we'll just be accessing that array at some arbitrary index currently pointed to by front(). This access is invalid as far the logic of front() is concerned but perfectly valid from an ASan point of view. The Microsoft STL with `DEBUG` defined include bounds-checking assertions that apparently aren't there in libc++.

As to the other part, I'm less sure. The third condition in the comment below was what prompted me to write it this way. Namely that if `scheduled_` is empty the `until` entry must have been removed. I can't see any benefit to continue waiting until the previously scheduled overdue-time for something else to be added, especially since we rarely (ever?) schedule things for execution in the past. That implies that when something new is added, it's not going to match the old `until` but it won't be overdue, so a new entry is going to cause us to break out, go to the top of the loop anyway so that we can reevaluate `until`.

 If there's nothing in the queue we might as well reset until the indefinite blocking above immediately.

Reply:
I should mention that I don't feel strongly about this one way or the other, but that I don't see how the change you're asking for gets us to "this thread only wakes up when there is a potentially overdue task"

Reply:
> the backing array will still exist

Makes sense, I thought about that.

> we rarely (ever?) schedule things for execution in the past.

This is currently [prohibited](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/util/executor_std.cc#L64-L67).

> I can't see any benefit to continue waiting until the previously scheduled overdue-time for something else to be added

SGTM. Do you think the comment would be clearer if this was mentioned explicitly?
```
// - `until` entry has been removed (including the case when the
//    queue became empty).
```

Reply:
It certainly doesn't hurt so: done.